File: aoc-2018/day-01/day_1.hs
import qualified Data.Set as Set

main = do
    content <- readFile "input1.txt"
    let numbers = getNumbers content
    putStr "Part 1: "
    print $ sum numbers
    putStr "Part 2: "
    print $ part2 numbers


-- Extract all the numbers from the input file text.
-- Standard read doesn't handle leading + signs.
getNumbers :: String -> [Int]
getNumbers content = map (read::String->Int) $ map (filter notPlusSign) $ lines content
    where
        notPlusSign x = x /= '+'

part2 :: [Int] -> Int
part2 numbers = solve numbers 0 (Set.fromList [0])

-- Find the first freq that is seen twice.
solve :: [Int] -> Int -> Set.Set Int -> Int
solve (x:xs) freq seen
    | Set.member nextTerm seen = nextTerm
    | otherwise = solve (xs ++ [x]) nextTerm $ Set.insert nextTerm seen
    where 
        nextTerm = freq + x

File: aoc-2019/day-01/day_1.py
import numpy as np

data = np.loadtxt('input.txt', dtype=int)

print("Part 1:", np.sum(data // 3 - 2))

total_fuel = 0

for mass in data:
    while True:
        fuel_needed = mass // 3 - 2

        if fuel_needed <= 0:
            break

        total_fuel += fuel_needed
        mass = fuel_needed


print("Part 2:", total_fuel)
